### ISSUE TITLE
WORKFLOW B — COMPLETE RL TRAINING PIPELINE IMPLEMENTATION

### ISSUE THREAD

**Issue by athariandre** [athariandre/yasta-bot#3]
Important:
This issue describes exactly what you must implement.
Do not deviate from this structure.
Do not introduce additional libraries without permission.
The code must be CPU-only and fully compatible with the tournament constraints.

The existing files from Workflow A include:

env.py

features.py

heuristic_bot.py

utils.py

Do NOT modify them, except to import from them.

Your job is to implement Workflow B, which includes:

✅ Behavior Cloning
✅ Self-Play PPO
✅ Residual CNN policy
✅ Replay Buffers
✅ Training Utilities
✅ Evaluation tools
✅ Final exportable agent.py for competition

Follow this spec line-by-line.

✅ 1. DIRECTORY LAYOUT (CREATE EXACTLY THIS)
workflow_b/
    policy.py
    bc_dataset.py
    bc_train.py
    selfplay_env.py
    buffer.py
    ppo_selfplay.py
    train_all.py
    eval.py
    make_agent_export.py
models/
    (auto generated .pth files)

Do not change directory names.

✅ 2. CORE MODEL — Residual CNN (AlphaZero-style)

Implement in workflow_b/policy.py

Architecture:

Input: (8, 18, 20) float tensor

Conv Stem:
    Conv2d(8 → 64, kernel=3, padding=1)
    BatchNorm2d(64)
    ReLU

Residual Tower:
    10 residual blocks
    Each block:
        Conv2d(64 → 64, kernel=3, padding=1)
        BatchNorm2d(64)
        ReLU
        Conv2d(64 → 64, kernel=3, padding=1)
        BatchNorm2d(64)
        Skip connection
        ReLU

Policy Head:
    Conv2d(64 → 4, kernel=1)
    Flatten
    Output: logits for 4 actions

Value Head:
    Conv2d(64 → 1, kernel=1)
    ReLU
    Flatten
    Linear → 64
    ReLU
    Linear → 1 (tanh)

Requirements:

Use PyTorch only.

Use CPU tensors only.

Must be deterministic (no dropout, no randomness).

Output a tuple: (policy_logits, value_scalar).

✅ 3. BEHAVIOR CLONING DATASET (bc_dataset.py)

Implement:

✅ Function: generate_bc_data(num_games, output_path)

This must:

Import env + heuristic bot

Simulate random games using the heuristic bot playing both sides

Collect training samples:

obs: (8,18,20)
action: integer 0–3

Store them in a dictionary:

{
    "obs": float32 numpy array [N, 8, 18, 20],
    "actions": int64 numpy array [N]
}

Use compression via .npz

Generate ~20k–40k samples total

Both perspectives must be included (p1 & p2)

✅ 4. BEHAVIOR CLONING TRAINING (bc_train.py)

Implement training code:

Inputs:

dataset .npz

training hyperparams

Training setup:

Cross entropy on policy head ONLY

Ignore value head during BC

Adam optimizer with lr=1e-3

Batch size = 128

20–40 epochs

Shuffle dataset each epoch

Outputs:

Save model to: models/policy_bc.pth

Logging:

Print loss every epoch

Print training accuracy

Save training curve plot as models/bc_loss_curve.png

✅ 5. SELF-PLAY ENVIRONMENT (selfplay_env.py)

Implement a wrapper for PPO self-play.

Class: SelfPlayWrapper

Responsibilities:

Maintain two policies:

current policy

opponent policy (frozen snapshot)

Load env from env.py

For each episode:

alternate perspectives (p1 or p2)

current policy plays as "me"

opponent policy plays "opp"

Return:

obs, action, reward, next_obs, done

Reward function:

If I survive and opponent dies → +1
If both die → 0
If I die → -1
Otherwise → 0 during the episode

Important:
No step-wise dense rewards, only terminal.

✅ 6. PPO REPLAY BUFFER (buffer.py)

Implement:

class PPOBuffer:
    store(obs, action, reward, value, logprob)
    finish_path(last_value)
    get()

Must compute:

GAE-Lambda

Advantages

Returns

Hyperparams:

gamma = 0.99
lambda = 0.95

Ensure:

normalize advantages

no device mismatch (CPU only)

✅ 7. PPO SELF-PLAY TRAINING (ppo_selfplay.py)

Implement:

PPO loop

Periodic opponent snapshotting

Policy update steps

Algorithm details:

Hyperparams:
steps_per_iteration = 8192
ppo_epochs = 4
mini_batch_size = 512
clip_ratio = 0.2
value_loss_coef = 0.5
entropy_coef = 0.01
lr = 3e-4
opponent_update_freq = 5 iterations
total_iterations ≈ 300–800 (long-running)

Behavior:

Generate experience via self-play

Compute advantages via buffer

Run PPO update

Every N iterations:

copy current policy into opponent snapshot

save snapshot under models/sp_archive/

Save model every iteration: models/policy_ppo_latest.pth

Stability requirements:

Clip gradients (torch.nn.utils.clip_grad_norm_)

Normalize observations? No (already binary channels)

Deterministic sampling using categorical(logits)

✅ 8. MASTER TRAINING SCRIPT (train_all.py)

This script should:

Generate dataset:

python bc_dataset.py

Run BC training:

python bc_train.py

Launch PPO self-play:

python ppo_selfplay.py

Display progress logs

Handle CTRL-C gracefully and save model on exit

✅ 9. EVALUATION SCRIPT (eval.py)

Implement:

Load any model file

Run X games vs:

heuristic bot

random bot

mirror (self)

Print win/loss/draw stats

Allow CLI arguments:

python eval.py --model models/policy_ppo_latest.pth --games 200

✅ 10. EXPORTER — CREATE SUBMISSION AGENT.PY (make_agent_export.py)

Implement:

This script must:

Load the chosen model (usually policy_ppo_latest.pth)

Convert the policy into a minimal inference-only version

Save single file: agent.py

This file must contain:

Model definition

Policy inference function

Minimal env integration framework

Requirements for final agent.py:

No training code

No self-play logic

No massive dependencies

CPU-only

Must run within tight time limits

Minimal imports:

import numpy as np
import torch

Must integrate with competition framework:

def agent(observation):
    return action_id

The exported agent must:

Load state_dict on first call

Perform forward pass with model.eval()

Select action = argmax(policy_logits)

✅ 11. GENERAL REQUIREMENTS

Code must be clean, modular, readable

Use Python type hints everywhere

Include docstrings for all public functions and classes

Avoid unnecessary classes

No external libraries beyond:

numpy

torch

matplotlib (for BC charts only)

Must run on CPU-only environment

Fully deterministic

✅ 12. COMPLETION CRITERIA

You are done when:

✅ BC dataset successfully generated
✅ BC model trains end-to-end
✅ PPO self-play runs for multiple iterations
✅ Policies evolve and improve
✅ eval.py prints meaningful win-rates
✅ make_agent_export.py outputs a working tournament agent.py
✅ All files exist exactly as specified
✅ Code runs with python train_all.py
